sections/implementation/recommender.tex

```diff
@@ -98,3 +98,133 @@ \subsection{Pros and cons}
+
+\begin{lstlisting}[caption={}, captionpos=b, label={}]
+    const alphaValue = 0.001;
+    const betaValue = 0.001;
+    for (let i = 0; i < numberOfRows; i++) {
+        for (let j = 0; j < numberOfCols; j++) {
+            if (userServiceMatrix[i + 1][j + 1] > 0) {
+                const error = userServiceMatrix[i + 1][j + 1] - predictedRatings[i][j];
+                for (let k = 0; k < numberOfFactors; k++) {
+                    userFactorMatrix[i][k] =
+                        userFactorMatrix[i][k] +
+                        alphaValue * (error * serviceFactorMatrix[k][j] - betaValue * userFactorMatrix[i][k]);
+                    serviceFactorMatrix[k][j] =
+                        serviceFactorMatrix[k][j] +
+                        alphaValue * (error * userFactorMatrix[i][k] - betaValue * serviceFactorMatrix[k][j]);
+                }
+            }
+        }
+    }
+    \dots
+\end{lstlisting}
+After updating the values, we calculate the dot product of the two factor matrices again, and use the new predictions to calculate the squared error.
+The squared error is calculated with the equation defined in \autoref{fig:minimizesquarederror}.
+We initially look at the matrix containing our actual ratings.
+We traverse it, and when the value of an entry is greater than 0, we know there is a rating.
+We subtract the prediction from the actual rating and square it, and then for the number of factors we add the regularization by squaring the values from both factor matrices, and multiplying it with the $\lambda$ value.
+\begin{lstlisting}[caption={}, captionpos=b, label={}]
+    let squareError = 0;
+    for (let i = 0; i < numberOfRows; i++) {
+        for (let j = 0; j < numberOfCols; j++) {
+            if (userServiceMatrix[i + 1][j + 1] > 0) {
+                squareError =
+                    squareError + Math.pow(userServiceMatrix[i + 1][j + 1] - predictedRatings[i][j], 2);
+                for (let k = 0; k < numberOfFactors; k++) {
+                    squareError =
+                        squareError +
+                        betaValue *
+                            (Math.pow(userFactorMatrix[i][k], 2) + Math.pow(serviceFactorMatrix[k][j], 2));
+                }
+            }
+        }
+    }
+    \dots
+\end{lstlisting}
+Once the maximum number of iterations has been performed, we need to make our data persist.
+To do this, we keep a separate ratings table in the database consisting of predicted ratings.
+To send the calculated ratings, we add them to the user by service matrix.
+Since we want to avoid possibly recommending a service the user has already rated, whenever we encounter an entry that is not 0, we remove that rating and assign it to be 0 instead.
+This means that, even if a user rated a service with the maximum value, they will not get it as a recommendation since it will now rank in the bottom with a rating of 0.
```

Comment:
Ikke så meget rettelse, men spørgsmål - kunne man ikke sætte den til -1 eller sådan noget, så det er åbenlyst at den ikke bare er en dårlig recommendation?

Reply:
Kan man godt, men det ændrer vel ikke så meget, skal bare sørge for at den ikke bliver valgt til top k